android/sdl_android/src/main/java/com/smartdevicelink/managers/lockscreen/LockScreenDeviceIconManager.java

```diff
@@ -0,0 +1,164 @@
+    }
+
+    void saveFileToCache(Bitmap icon, String iconUrl) {
+
+        String iconHash = getMD5HashFromIconUrl(iconUrl);
+
+        File f = new File(this.context.getCacheDir() + "/" + STORED_ICON_PATH, iconHash);
+        ByteArrayOutputStream bos = new ByteArrayOutputStream();
+        icon.compress(Bitmap.CompressFormat.PNG, 0 /*ignored for PNG*/, bos);
+        byte[] bitmapdata = bos.toByteArray();
+
+        FileOutputStream fos = null;
+        try {
+            fos = new FileOutputStream(f);
+            fos.write(bitmapdata);
+            fos.flush();
+            fos.close();
+        } catch (Exception e) {
+            DebugTool.logError("Failed to save icon to cache");
+            e.printStackTrace();
+            return;
+        }
+
+        JSONObject iconParams;
+        try {
+            iconParams = buildDeviceIconParameters(f.getAbsolutePath());
+            writeDeviceIconParametersToSystemPreferences(iconHash, iconParams);
+        } catch (JSONException e) {
+            DebugTool.logError("Failed to save to shared preferences, clearing cache icon directory");
+            clearIconDirectory();
+            e.printStackTrace();
+        }
+    }
+
+    Bitmap getFileFromCache(String iconUrl) {
+        String iconHash = getMD5HashFromIconUrl(iconUrl);
+        SharedPreferences sharedPref = this.context.getSharedPreferences(SDL_DEVICE_STATUS_SHARED_PREFS, Context.MODE_PRIVATE);
+        String iconParameters = sharedPref.getString(iconHash, null);
+
+        if (iconParameters != null) {
+            JSONObject jsonObject = null;
+            try {
+                jsonObject = new JSONObject(iconParameters);
+                String storedUrl = jsonObject.getString(STORED_URL);
+                Bitmap cachedIcon = BitmapFactory.decodeFile(storedUrl);
+                if(cachedIcon == null) {
+                    DebugTool.logError("Failed to get Bitmap from decoding file cache");
+                    clearIconDirectory();
+                    return null;
+                } else {
+                    return cachedIcon;
+                }
+            } catch (JSONException e) {
+                DebugTool.logError("Failed to get file from cache, removing shared pref");
+                sharedPref.edit().remove(iconHash).commit();
+                e.printStackTrace();
+                return null;
+            }
+        } else {
+            DebugTool.logError("Failed to get system preferences");
+            return null;
+        }
+    }
+
+    private void writeDeviceIconParametersToSystemPreferences(String iconHash, JSONObject jsonObject) {
+        SharedPreferences sharedPref = this.context.getSharedPreferences(SDL_DEVICE_STATUS_SHARED_PREFS, Context.MODE_PRIVATE);
+        SharedPreferences.Editor editor = sharedPref.edit();
+        editor.putString(iconHash, jsonObject.toString());
+        editor.commit();
+    }
+
+    private JSONObject buildDeviceIconParameters(String storedUrl) throws JSONException {
```

Comment:
Do you think we can simplify that a bit by only storing the hash as key and the time as a value instead of saving the JSON in shared prefs?

currently, this is how the share prefs xml look like. We save the hash twice. I feel that is unesceasry 
```
<?xml version='1.0' encoding='utf-8' standalone='yes' ?>
<map>
    <string name="935e06761f887b20a5a10242faaf6ecf">{&quot;storedUrl&quot;:&quot;\/data\/user\/0\/io.livio.sdltestsuite.app1\/cache\/sdl\/lock_screen_icon\/935e06761f887b20a5a10242faaf6ecf&quot;,&quot;lastUpdatedTime&quot;:1585600940171}</string>
</map>
```

Cannot we just store them like that? This way we don't even need to make s JSON.
```
<?xml version='1.0' encoding='utf-8' standalone='yes' ?>
<map>
    <string name="935e06761f887b20a5a10242faaf6ecf">1585600940171</string>
</map>
```

Reply:
Implemented suggestion and fixed associated code